guidance/library/_json.py

```diff
@@ -77,47 +71,84 @@ def json(
             - A subclass of ``pydantic.BaseModel``
             - An instance of ``pydantic.TypeAdapter``
     """
-    if "compact" in kwargs:
-        warnings.warn("The 'compact' argument is deprecated and has no effect. It will be removed in a future release.", category=DeprecationWarning)
-        kwargs.pop("compact")
-    if kwargs:
-        raise TypeError(f"Unexpected keyword arguments: {kwargs.keys()}")
-    if schema is None:
+    if schema is False:
+        raise ValueError("Unsatisfiable schema: schema is false")
+    elif schema is True:
+        schema = {}
+    elif schema is None:
         # Default schema is empty, "anything goes" schema
         # TODO: consider default being `{"type": "object"}`
         schema = {}
-    elif isinstance(schema, (Mapping, bool, str)):
-        if isinstance(schema, str):
-            schema = cast(JSONSchema, json_loads(schema))
-    elif isinstance(schema, pydantic.TypeAdapter) or (isinstance(schema, type) and issubclass(schema, pydantic.BaseModel)):
+    elif isinstance(schema, pydantic.TypeAdapter) or (
+        isinstance(schema, type) and issubclass(schema, pydantic.BaseModel)
+    ):
         schema = pydantic_to_json_schema(schema)
+    elif isinstance(schema, str):
+        from_str = json_loads(schema)
+        if not isinstance(from_str, dict):
+            raise ValueError("JSON schema string must be a JSON object (i.e. a dictionary)")
+        schema = from_str
+
+    if isinstance(schema, Mapping):
+        schema = dict(schema)
     else:
         raise TypeError(f"Unsupported schema type: {type(schema)}")
 
-    compiler = JsonCompiler(
-        separators=separators,
-        whitespace_flexible=whitespace_flexible,
-        coerce_one_of=False,
-    )
-
-    schema_string = json_dumps(schema)
-    try:
-        llg = compiler.compile(schema_string)
-    except ValueError as e:
-        if e.args[0] == "oneOf constraints are not supported. Enable 'coerce_one_of' option to approximate oneOf with anyOf":
-            warnings.warn("oneOf not fully supported, falling back to anyOf. This may cause validation errors in some cases.")
+    coerce_one_of = False
+    # TODO: decide whether or not to keep this -- it lets us double check that llguidance can handle the schema (which isn't necessarily)
+    # what we want, as llguidance may or may not be the backend we are using. That being said, it's sort of nice to get an exception when
+    # you call `json` instead of waiting for generation to fail.
+    VALIDATE = True
+    if VALIDATE:
+        schema_string = json_dumps(schema)
+        try:
             compiler = JsonCompiler(
                 separators=separators,
                 whitespace_flexible=whitespace_flexible,
-                coerce_one_of=True,
+                coerce_one_of=False,
             )
-            llg = compiler.compile(schema_string)
+            compiler.compile(schema_string)
```

Comment:
note to self: add compiler.check() in llguidance so we don't have to expose the JSON grammar format externally